### ISSUE TITLE
File Uploads within Array Public Property

### ISSUE THREAD

**Issue by jjjrmy** [livewire/livewire#1222]
When using file-upload and wire model as an array, there are some issues on hydration where the data is lost.
![Help](https://i.imgur.com/j39w5C1.gif)

```php
<?php

namespace App\Http\Components;

use Livewire\Component;
use Livewire\WithFileUploads;

class Form extends Component
{
    use WithFileUploads;
    public $a = [];

    public function mount() {
        $this->a['file_upload'] = []; // Get Input Ready For MultiFile
    }
}
```

```html
<input type="text" name="first_name" wire:model.lazy="a.first_name" />
<input type="file" multiple name="files" wire:model="a.file_upload" />

<?php dump($this->a); ?>
```

**Comment by jkque** [comment]
Got this working by tweaking the `WithFileUploads.php` .

```php
    protected function dehydratePropertyFromWithFileUploads($name, $value)
    {
        if ($value instanceof TemporaryUploadedFile) {
            return $value->serializeForLivewireResponse();
        } elseif (is_array($value) && isset($value[0]) && $value[0] instanceof TemporaryUploadedFile) {
            return $value[0]::serializeMultipleForLivewireResponse($value);
        } elseif (is_array($value) && $this->isAssoc($value)) {
            $key = collect(array_keys($value))->map(function ($key) use ($value) {
                if ($value[$key] instanceof TemporaryUploadedFile) {
                    return [$key, 'singleUpload'];
                } elseif (is_array($value[$key]) && isset($value[$key][0]) && $value[$key][0] instanceof TemporaryUploadedFile) {
                    return [$key, 'multipleUpload'];
                }
            })->first();

            if($key) {
                if ($key[1] === 'singleUpload') {
                    $value[$key[0]] = $value[$key[0]]->serializeForLivewireResponse();
                }

                if ($key[1] === 'multipleUpload') {
                    $value[$key[0]] = $value[$key[0]][0]::serializeMultipleForLivewireResponse($value[$key[0]]);
                }
            }
        }

        return $value;
    }

    protected function isAssoc(array $arr) 
    {
        if (array() === $arr) return false;
        return array_keys($arr) !== range(0, count($arr) - 1);
    }
```

added 2nd elseif checking if the value is an associative array

**Comment by mblarsen** [comment]
Can you use the diff tag for the code block above to better illustrate your changes:

```diff
- removed
  unchanged
+ added
```

Like so:

    ```diff
    - removed
      unchanged
    + added
    ```

**Comment by jkque** [comment]
edited my comment, sorry i am bad at formatting comments

**Comment by calebporzio** [comment]
Heads up: There's a $200 bug bounty on this issue: https://www.bountysource.com/issues/91956429-file-uploads-within-array-public-property

**Comment by mallardduck** [comment]
@jkque - you should hop on that bounty! looks like you've done most the work for it. 😄

**Comment by jkque** [comment]
Don't know how to submit a PR. First time to contribute on open source project.

**Comment by Te7a-Houdini** [comment]
@jkque [check these steps](https://opensource.com/article/19/7/create-pull-request-github) may be helpful to make your first time contribution :wink:

**Comment by iantasker** [comment]
I have created a PR with a test to fix the issue but happy to defer to @jkque if they can create a PR as they had a working version first.

**Comment by jjjrmy** [comment]
@iantasker @jkque 

It is still not working and my guess is it's related to this line

![Not working](https://i.imgur.com/qduMvFE.gif)

**Comment by iantasker** [comment]
@jjjrmy  thanks for letting me know. I will try and replicate with surge.

**Comment by iantasker** [comment]
@jjjrmy I have been able to replicate the issue in surge (https://github.com/livewire/surge/commit/162523b7a9f81728f79b92e8c91584ed6ba9fcca) and have updated the PR with the required changes.

**Comment by jjjrmy** [comment]
@iantasker Still not working, now the input data is missing

![broken](https://i.imgur.com/249GBLu.gif)

**Comment by jkque** [comment]
@Te7a-Houdini - thanks dude, just submitted my first PR 

@jjjrmy - already submitted a pr with a fixed of  @iantasker solution

**Comment by jjjrmy** [comment]
@jkque Checking it now, I think it's working.

**Comment by jkque** [comment]
@jjjrmy - does it work in your end ?

**Comment by jjjrmy** [comment]
@jkque It looks like it's working but when it rehydrates the `basename` and `pathname` are not the same. Is that expected behavior? I just started with the upload feature so not sure if that should change every hydrate.

**Comment by jkque** [comment]
@jjjrmy - i think that is the expected behavior

**Comment by Rakesh717** [comment]
> @jjjrmy - i think that is the expected behavior

If it is the expected behavior then this code would be more cleaner:
```
   protected function dehydratePropertyFromWithFileUploads($name, $value)
    {
        if ($value instanceof TemporaryUploadedFile) {
            return $value->serializeForLivewireResponse();
        } elseif ($this->isAssoc($value)) {
            foreach (array_keys($value) as $key) {
                if ($value[$key] instanceof TemporaryUploadedFile) {
                    $value[$key] = $value[$key]::serializeForLivewireResponse();
                } elseif (is_array($value[$key]) && isset($value[$key][0]) && $value[$key][0] instanceof TemporaryUploadedFile) {
                    $value[$key] = $value[$key][0]::serializeMultipleForLivewireResponse($value[$key]);
                }
            }
        }
        return $value;
    }

    protected function isAssoc(array $arr)
    {
        if (array() === $arr) return false;
        return array_keys($arr) !== range(0, count($arr) - 1);
    }
```

**Comment by iantasker** [comment]
@jjjrmy I have updated my PR to fix the issue and also include an explicit test to check additional properties are retained.

**Comment by jjjrmy** [comment]
Issue fixed thanks to #1234
